<a href="https://colab.research.google.com/github/hmayorgav/discretas/blob/main/taller.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## **Taller - teoría de números - HAIDER ANDRES MAYORGA**

**Matemáticas Discretas 2: 2023-1**


**Universidad Nacional De Colombia**

### **1. ¿Existen enteros $a$ y $b$ tal que $a + b = 544$ y cuyo máximo común divisor es 11?.**

Para comprobar la existencia de tal enunciado, es necesario conocer un numero a que sean multiplos de 11, con ello revisar un numero b que tambien sean divisibles por 11 pero tiene que ser resultante de 544-11*a=b, con ello evaluamos el mcd con ayuda del algoritmo de euclides

In [ ]:
def mcd(a, b):
    if b==0:
        return a
    return mcd(b,a%b)
for a in range(0,544,11):
  b=544-a
  if(mcd(a,b)==11):
     print(a,b) 


Como es posible observar no existe dos numeros a,b que se cumpla 544 ≡ a+b (mod 11)  

### **2. Encuentre una regla de divisibilidad para 8 y para 16.**

Regla de divisibilidad del 8
-tomar las ultimas 3 cifras, Si estas son divisibles entre 8, entonces son divisibles entre 8

Regla de divisibilidad del 16
-tomar ultimas 4 cifras, si estas son divisibles entre 16, entonces son divibles entre 16



### **3. Si $p$ es un número primo y $a^2 \equiv b^2 (\mod p)$, pruebe que $a \equiv \pm b$.**


Inicialmente tenemos que $$a^2 \equiv b^2 (\mod p) \implies (a−b)(a+b)≡0(mod \ p)$$entonces puede ser expresado como $p|(a-b)(a+b)$, pero si P es primo por el lema de euclides que menciona que si $$n|ab,mcd(n,a)=1 \implies n|b $$ que por tanto  $p|a+b $ y $p|a-b $ que puede re-escrito como $a \equiv b (\mod p)$ y $a \equiv -b (\mod p)$

















### **4. Encuentre el resto cuando $19^{19}$ es dividido por 5.**

partiendo de $$Si \ a \ y \ n \ son \ enteros  \ primos \ relativos, \ entonces 
 \ a^{φ(n)} ≡ 1 (mod \ n)$$
ademas de 
$$para \ primos \ p, φ(p)=p-1$$
podemos dividir a $19^{19}$ como $19^{4}$*$19^{4}$*$19^{4}$*$19^{4}$*$19^{1}$
por tanto al aplicar el teorema de euler con mod 5(numero primo) entonces el problema se reduce a  $19^{1}(mod \ 5) \implies 4$

### **5. Encuentre los últimos dos dígitos de $7^{7^{7}}$.**


partiendo de reducir el problema a $$7^{7^7} \equiv 7^{\alpha}(mod \ 100)$$
por tanto 
$$7^{7} \equiv \alpha(mod \ \phi(100))$$
podemos por propiedades de $phi$
$$\phi(100)=\phi(25)\phi(4)$$
$$\phi(100)=\phi(5^2)\phi(2^2)$$
$$\phi(100)=2^1(2-1)5^1(5-1)$$
$$\phi(100)=40$$
seguimos reduciendo el mod 

$$7^{7} \equiv \alpha(mod \ \phi(40))$$
$$\phi(40)=\phi(2^3)\phi(5)$$
$$\phi(40)=2^2(2-1)4$$
$$\phi(40)=16$$

ahora que tenemos una version simplificada podemos dar solucion a 
$$7^{7} \equiv \alpha(mod \ 16)$$
$$7^{2}7^{2}7^{2}7 \equiv \alpha(mod \ 16)$$
$$1*1*1*7 \equiv \alpha(mod \ 16)$$
$$7 \equiv \alpha(mod \ 16)$$

volvemos a nuestro problema original 
$$7^{7} \equiv 7^{\alpha}(mod \ 100)$$
$$7^{7} \equiv x(mod \ 100)$$
$$7^{3}7^{3}7^{1} \equiv x(mod \ 100)$$
recordando que $7^3=343$ y que $343 mod(100)=43$
$$43 \ 43 \ 7^{1} \equiv x(mod \ 100)$$
recordando que $43*7=301$ y que $301 mod(100)=1$
$$ 43*1\equiv x(mod \ 100)$$
$$43\equiv x(mod \ 100)$$
$$7^{7^7} (mod \ 100)=43$$

### **6. Encuentre $\phi(n)$ para n=35,  n=100, n=51200.**

Para este podemos hacer uso del algoritmo ya creado
por este link 
[phy funtion code](https://github.com/hmayorgav/discretas/blob/main/13%20mar%20Jupyter%202%20comprobar%20el%20tiempo%20por%20el%20m%C3%A9todo%20tradicional%20de%20comprobar%201%20a%201%2C%20vs.%20Totient%20de%20Euler/13mar%20Totient%20de%20Euler.ipynb)


n=35 ϕ(n)=24

n=100 ϕ(n)=40

n=51200 ϕ(n)=20480


### **7. Usted le pregunta a un robot que quiere comer. El responde “48.879”. Sabiendo que el robot piensa en hexadecimal pero habla el decimal, que le debería dar de comer?**

para resolver este problema necesitamos dividir el numero entre 16 y con el residuo volver a dividir entre 16, esto hasta que el residuo sea cero y imprima la traduccion en hexadecimal que seran los residuos en inversa, lo cual se puede realizar con el siguiente codigo

In [ ]:
num= 48879
word=""
while(True):
  res=num%16
  if(res<10):
    word+=str(res)
  elif(res==10):
     word+="a"
  elif(res==11):
     word+="b"
  elif(res==12):
    word+="c"
  elif(res==13):
    word+="d"
  elif(res==14):
     word+="e"
  elif(res==15):
     word+="f"
  if(num//16==0):
    break
  else:
    num//=16
print(word[::-1])

beef


### **8. ¿65.314.638.792 es divisible por 24?.**

si es divisible entre 8 y 3 por tanto las reglas de divisibilidad

para el 3 sera que la suma de sus cifras sea multiplo de 3 por tanto
$$6+5+3+1+4+6+3+8+7+9+2 mod(3)$$
$$54 \  mod(3)=0$$
para el 8 sera que sus ultimos 3 digitos juntos sean multiplo de 8, por tanto
$$792 mod(8)=0$$
esto se puede debido a que 
$$mcd(3,8)=1 \ y \ 24 \ cumple \ que \ 3|24, \ 8|24 \implies 3\ 8|24 $$ 



### **9.Pruebe que $n^p - n$ es divisible por $p$ si $p$ es un número primo.**

se puede ver como un caso especial del teorema de euler.
Si a,p son coprimos, entonces:
$a^{\phi(p)}  \equiv  1 \mod(p)$
entonces, sabiendo que p numero es primo $ \phi(p)=1 $ conseguimos la siguiente expresion $a^{\phi(p)}  \equiv  1 \mod(p)$, por tanto al utilizar la identidad de tomando dos numeros a,b tal que  $a \equiv b \mod(n)$, entonces con cualquier numero c  $ac \equiv bc \mod(n)$, obtenemos la siguiente expresion multiplicando a a cada lado de la expresion, por tanto $a^p \equiv a \mod(n)$
que se puede expresar como $a^p - a\equiv 0 \mod(n)$ o $p|a^p - a$ 

###**10. Encuentre los enteros $x$ y $y$ tal que $314x+159y=1$.**###


conociendo inicialmente que a traves del algoritmo de euclides se sabe que

In [5]:
def mcd(a, b):
    if b==0:
        return a
    return mcd(b,a%b)
mcd(314,159)

1

podemos hacer uso de la identidad de bezout que indica,
para todo a y b, existen un x y un y tales que
$$ax+by=mcd(a,b)$$

a continuacion la implementacion de la identidad de bezout para encontrar la solucion al problema

In [6]:

def bezout(a, b):
    if b == 0:
        return 1, 0, a
    else:
        x, y, gcd = bezout(b, a % b)
        return y, x - (a // b) * y, gcd
print(bezout(314,159))


(-40, 79, 1)


por tanto la solucion al sistema es $x=79 \ y \ y=-40$ 

### **11. Pruebe o controvierta la siguiente afirmación si $a^2 \equiv b^2 (\mod m)$ entonces $a \equiv b (\mod m)$ o $a \equiv -b (\mod m)$.**

se puede demostrar por contraejemplo el enunciado anterior, sea a=4 \ b=16 \ y \ m=15
se puede encontrar que 
$$4^2 \equiv 16^2 (\mod 15)$$
$$16 \equiv 256 (\mod 15)$$
que es cierto,ahora probamos la consecuencia
$$4 \equiv 16 (\mod 15)$$
o
$$4 \equiv -16 (\mod 15)$$
lo cual es falso, por tanto la afirmacion es falsa.

### **12. Encuentre todos los enteros positivos tales que $1066 \equiv 1776 (\mod m)$**






Para encontrar la respuesta podemos inicialmente convertir la expresion a $710 \equiv 0 (\mod m)$, ahora bien la pregunta se convierte en cuales son los divisores de 710, los cuales se pueden hallar con este pequeño pedazo de codigo



In [11]:
import math
def divisores(a):
  divisores=[]
  for i in range(1,a+1):
    if(a%i==0):
      divisores.append(i)
  
  return divisores
print("los divisores son",divisores(710))

los divisores son [1, 2, 5, 10, 71, 142, 355, 710]


### **13. Muestre que la diferencia de dos cubos consecutivos nunca es divisible por 5.**

Supongamos que $n$ y $n+1$ son dos números consecutivos. Queremos demostrar que la diferencia de los cubos de estos dos números, es decir, $n^3 - (n+1)^3$, no es divisible por 5.

Podemos escribir $n^3 - (n+1)^3$ como:

\begin{align*}
(n+1)^3 - (n)^3 &= n^3+3n^2+3n+1-n^3\\
&= 3n(n+1)+1\\
\end{align*}

Ahora,pensemos en que probando con las formas de expresar tal numero que son $5*n,5*n+1,5*n+2,5*n+3,5*n+4$, por tanto probando manualmente cada caso podemos revisar la afirmacion
al revisar caso por caso, podemos concluir con lo siguiente
 1. 5n
$$-3(5n)(5n+1)-1$$
$$u=-3(5n+1)$$
$$5u-1$$
osea
$$1\not\equiv 0(\mod 5)$$
 2. 5n+1
$$3(5n+1)(5n+2)+1$$
$$5(3)(5n^2+2n+n)+7$$
$$u=(3)(5n^2+2n+n)$$
$$5u+7$$
osea $$2\not\equiv 0(\mod 5)$$
 
 3. 5n+2
$$3(5n+2)(5n+3)+1$$
$$5(3)(5n^2+3n+2n)+19$$
$$u=(3)(5n^2+3n+2n)$$
$$5u+19$$
osea $$4\not\equiv 0(\mod 5)$$

 4. 5n+3
$$3(5n+3)(5n+4)+1$$
$$5(3)(5n^2+4n+3n)+37$$
$$u=(3)(5n^2+4n+3n)$$
$$5u+37$$
osea $$2\not\equiv 0(\mod 5)$$
 5. 5n+4
$$3(5n+4)(5n+5)+1$$
$$5(3)(5n^2+5n+4n)+61$$
$$u=(3)(5n^2+5n+4n)$$
$$5u+61$$
osea $$1\not\equiv 0(\mod 5)$$
al ninguno ser congruente con 5, se concluye que no existe diferencia de cubos que sea congruente con 5

### **14. Encuentre un entero positivo $n$ tal que $3^2|n$, $4^2|n+1$, $5^2|n+2$**

el problema se puede expresar como 
$$n \equiv 0(\mod 3^2)$$
$$n \equiv -1 (\mod 4^2)$$
$$n \equiv -2 (\mod 5^2)$$
teniendo en cuenta que entre si todos son coprimos se puede solucionar por teorema de resto chino
$$m=9*16*25=3600$$
$$M_1=16*25=400$$
$$M_2=9*25=225$$
$$M_3=9*16=144$$
resolvemos los siguientes sistemas
$$400x \equiv 1 mod (3^2)$$
$$225x \equiv 1 mod (4^2)$$
$$144x \equiv 1 mod (5^2)$$
con ayuda de la identidad de Bezout



In [10]:
def bezout(a, b):
    if b == 0:
        return 1, 0, a
    else:
        x, y, gcd = bezout(b, a % b)
        return y, x - (a // b) * y, gcd
print("sol primer sistema",bezout(400,9)[0])
print("sol segundo sistema",bezout(255,9)[0])
print("sol tercer sistema",bezout(144,25)[0])

sol primer sistema -2
sol segundo sistema 1
sol tercer sistema 4


ahora tenemos que 
$$n=(-1)(9*25)(1)+(-2)(16*9)(4)$$
$$n=-1377$$
$$-1377 \  mod(m)$$
$$m=3600$$
$$-1377 \ mod(3600)=2223$$


### **15. ¿Cuál es el último digito de $7^{355}$?**

$7^{355} (mod 10) \equiv 7 7^{354} (mod 10) \equiv  (7^{2})^{177} (mod 10) \equiv  (49)^{177} (mod 10) \equiv  \equiv  7(-1)^{177} (mod 10) \equiv  -7 (mod 10) \equiv  -7 (mod 10)\equiv  3 (mod 10) $

### **16. Muestre que $3k+4$ y $4k+5$ no tienen un factor común más grande que 1**

Haciendo uso del algoritmo de euclides para hallar el mcd, como el siguiente


In [ ]:
def mcd(a, b):
    if b==0:
        return a
    return mcd(b,a%b)

podemos encontrar que 
 $$siendo \ a= 4k+5  \ y \ b= 3k+4$$
pasados como parametros tal que
 $$ a'=3k+4 \ y \ a \% b=k+1$$
 ahora bien
 $$ a''=k+1 \ y \ a'\%b=1 $$
$$ a'''=1 \ y \ a''\% \ b=0$$
como $a'''=1$ su mayor factor comun es 1